comms/src/inbound_message_service/duplicate_msg_cache.rs

```diff
@@ -0,0 +1,117 @@
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::consts::{DHT_MSG_CACHE_STORAGE_CAPACITY, DHT_MSG_CACHE_TTL};
+use derive_error::Error;
+use std::{hash::Hash, time::Duration};
+use ttl_cache::TtlCache;
+
+#[derive(Debug, Error)]
+pub enum DuplicateMsgCacheError {
+    /// A duplicate entry existed in the message cache
+    DuplicateEntry,
+}
+
+#[derive(Clone, Copy)]
+pub struct DuplicateMsgCacheConfig {
+    /// The maximum number of messages that can be tracked using the DuplicateMessageCache
+    storage_capacity: usize,
+    /// The Time-to-live for each stored message
+    msg_ttl: Duration,
+}
+
+impl Default for DuplicateMsgCacheConfig {
+    fn default() -> Self {
+        DuplicateMsgCacheConfig {
+            storage_capacity: DHT_MSG_CACHE_STORAGE_CAPACITY,
+            msg_ttl: DHT_MSG_CACHE_TTL,
+        }
+    }
+}
+
+/// The DuplicateMsgCache is used to track handled messages to ensure that processing resources are not wasted on
+/// duplicate messages and that these duplicate messages are not sent to services or propagate through the network.
+pub struct DuplicateMsgCache<K: Eq + Hash> {
+    config: DuplicateMsgCacheConfig,
+    cache: TtlCache<K, ()>,
+}
+
+impl<K> DuplicateMsgCache<K>
+where K: Eq + Hash
+{
+    /// Create a new DuplicateMsgCache with the specified configuration
+    pub fn new(config: DuplicateMsgCacheConfig) -> Self {
+        Self {
+            config,
+            cache: TtlCache::new(config.storage_capacity),
+        }
+    }
+
+    /// Insert a new message into the DuplicateMsgCache with a time-to-live starting at the insertion time. It will
+    /// return a DuplicateEntry Error if the message has already been added into the cache.
+    pub fn insert(&mut self, msg: K) -> Result<(), DuplicateMsgCacheError> {
+        match self.cache.insert(msg, (), self.config.msg_ttl) {
+            Some(_) => Err(DuplicateMsgCacheError::DuplicateEntry),
+            None => Ok(()),
+        }
+    }
+
+    /// Check if the message is available in the DuplicateMsgCache
+    pub fn contains(&self, msg: &K) -> bool {
+        self.cache.contains_key(msg)
+    }
+}
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::{thread, time::Duration};
+
+    #[test]
+    fn test_msg_rlu_and_ttl() {
+        let mut duplicate_msg_cache: DuplicateMsgCache<String> = DuplicateMsgCache::new(DuplicateMsgCacheConfig {
+            storage_capacity: 3,
+            msg_ttl: Duration::from_secs(1),
```

Comment:
The idea is 💯 , but can you reduce the times here by ±90% to keep the test snappy?

Reply:
Reduced the waiting time to 100ms